### PR TITLE
Frame control

### PR DESCRIPTION
**Pull request by trueromanus**
Hi! I want to share a small control that was inspired by a control with the same name from UWP ([link](https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.frame)).  
This control supports navigation to new pages, keep a navigation history and can perform step back and forward.  
Working [demo](https://github.com/trueromanus/AvaloniaUI.Navigation.Example) (only as source code, sorry for it).

### CODE REVIEW

File: src/Avalonia.Controls/Navigation/Page.cs [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+        }
+
+        /// <summary>
+        /// Invoked after load this the page in the frame.
+        /// </summary>
+        /// <param name="eventArguments">Class contains data related with navigation.</param>
+        public virtual void NavigateTo ( NavigationEventArgs eventArguments )
+        {
+        }
+
```
Comment (Gillibald) [inline src/Avalonia.Controls/Navigation/Page.cs]: You should probably add NavigatingTo() to cancel a Navigation or to prepare stuff before the frame displays the page.

File: src/Avalonia.Controls/Navigation/Page.cs [hunk 2]
```diff
@@ -0,0 +1,32 @@
...
+        /// <summary>
+        /// Frame that page belongs to.
+        /// </summary>
+        public Frame Frame { get; set; }
+
+        /// <summary>
+        /// Invoked when the page leave the frame.
+        /// </summary>
+        /// <param name="eventArguments">Class contains data related with navigation.</param>
+        public virtual void NavigateFrom ( NavigationEventArgs eventArguments )
```
Comment (Gillibald) [inline src/Avalonia.Controls/Navigation/Page.cs]: It should be named NavigatedFrom

File: src/Avalonia.Controls/Navigation/Frame.cs [hunk 3]
```diff
@@ -0,0 +1,290 @@
...
+using System.Collections.Generic;
+using System.Threading.Tasks;
+
+namespace Avalonia.Controls {
+
+    /// <summary>
+    /// Control allow execute navigation wihtin itself.
+    /// </summary>
+    public class Frame : Panel
+    {
```
Comment (Gillibald) [inline src/Avalonia.Controls/Navigation/Frame.cs]: Why did you choose a Panel? I dont think a Frame will have multiple childs. In my opinion ContentControl would be the better fit.

File: src/Avalonia.Controls/Navigation/Frame.cs [hunk 4]
```diff
@@ -0,0 +1,290 @@
...
+
+                _currentState = historyItem;
+                RaisePropertyChanged ( CurrentStateProperty , CurrentState , historyItem );
+            }
+        }
+
+        /// <summary>
+        /// This parameter influence on 
+        /// </summary>
+        public bool IsStack
```
Comment (Gillibald) [inline src/Avalonia.Controls/Navigation/Frame.cs]: Why dont you introduce a cache for used pages and let us set the cache size. By default the size could be one.

**Comment by Gillibald** [comment]
Does Avaloina have a common navigation strategy? Maybe we need some kind of navigation Service.

**Review by Gillibald — COMMENTED** [review]
Just left you a few comments. How do you handle errors? I think we need some events for the navigation process (NavigatingTo, NavigatedTo, NavigatedFrom)

**Comment by Gillibald** [comment]
I much prefer to use my own navigation in my projects. Build in navigation is always not flexible enough i always want to have control about navigation in a MVVM friendly way and thats hard to achieve with build in controls. You are probably right that a Frame Control should not be part of the core project.

**Comment by lindexi** [comment]
Thanks for this @trueromanus ! But your code format dont like Avalonia's code format, for space and the new line.

The suggestion is that you can install resharper and dont set any config and use `clean up code` to clean or use `ctrl`+`alt`+`enter` to format the code.

**Comment by danwalmsley** [comment]
@grokys Perhaps a control like this might become more important once our mobile backends are in a better state. dont UWP, Windows phone had templates where they had some kind of Frame control or navigation service?

**Comment by trueromanus** [comment]
Sorry for long delay with answer. I have issue with internet.
@grokys I thinked what need default solution out of box. There are button, grid, image, what wrong with Navigation control? :)
I do not insist @grokys @Gillibald if you think that this feature not for core it is ok, this is your project, and you know how it should be.
@grokys your opinion make sense and I think too what need another repo for this code part. I will to do another package for nuget.
@Gillibald you right, navigation service from Prism is very good (really) but for any control out of box can say what "not flexible enough". 
Thanks all.

**Comment by danwalmsley** [comment]
@grokys maybe we should think about some Avalonia support libraries things like this could be added to?

**Pull request closed without merging**